pkg/sql/catalog/descs/descriptor.go

```diff
@@ -230,137 +254,233 @@ func (q *byIDLookupContext) lookupLeased(
+	flags tree.CommonLookupFlags,
+	requestedType catalog.DescriptorType,
+) (catalog.Descriptor, error) {
+	mustBeVirtual, vd, err := tc.getVirtualDescriptorByName(sc, name, flags.RequireMutable, requestedType)
+	if mustBeVirtual || vd != nil || err != nil || (db == nil && sc != nil) {
+		return vd, err
 	}
-	return nil
+	id, err := tc.getNonVirtualDescriptorID(ctx, txn, db, sc, name, flags)
+	if err != nil || id == descpb.InvalidID {
+		return nil, err
+	}
+	// When looking up descriptors by name, then descriptors in the adding state
+	// must be uncommitted to be visible (among other things).
+	flags.AvoidCommittedAdding = true
+	descs, err := tc.getDescriptorsByID(ctx, txn, flags, id)
+	if err != nil {
+		// Swallow error if the descriptor is dropped.
+		if errors.Is(err, catalog.ErrDescriptorDropped) {
+			return nil, nil
+		}
+		if errors.Is(err, catalog.ErrDescriptorNotFound) {
+			// Special case for temporary schemas, which can't always be resolved by
+			// ID alone.
+			if db != nil && sc == nil && isTemporarySchema(name) {
+				return schemadesc.NewTemporarySchema(name, id, db.GetID()), nil
+			}
+			// In all other cases, having an ID should imply having a descriptor.
+			return nil, errors.WithAssertionFailure(err)
+		}
+		return nil, err
+	}
+	desc := descs[0]
+	if desc.GetName() != name && !(desc.DescriptorType() == catalog.Schema && isTemporarySchema(name)) {
+		// TODO(postamar): make Collection aware of name ops
+		return nil, nil
+	}
+	return desc, nil
 }
 
-func (tc *Collection) getByName(
+type continueOrHalt bool
+
+const (
+	continueLookups continueOrHalt = false
+	haltLookups     continueOrHalt = true
+)
+
+// getVirtualDescriptorByName looks up a virtual descriptor by name.
+//
+// Virtual descriptors do not always have an ID set, so they need to be treated
+// separately from getNonVirtualDescriptorID. Also, validation, type hydration
+// and state filtering are irrelevant here.
+func (tc *Collection) getVirtualDescriptorByName(
+	sc catalog.SchemaDescriptor,
+	name string,
+	isMutableRequired bool,
+	requestedType catalog.DescriptorType,
+) (continueOrHalt, catalog.Descriptor, error) {
+	objFlags := tree.ObjectLookupFlags{
+		CommonLookupFlags: tree.CommonLookupFlags{
+			RequireMutable: isMutableRequired,
+		},
+	}
+	switch requestedType {
+	case catalog.Schema:
+		if vs := tc.virtual.getSchemaByName(name); vs != nil {
+			return haltLookups, vs, nil
+		}
+	case catalog.Type:
+		objFlags.DesiredObjectKind = tree.TypeObject
+		fallthrough
+	case catalog.Table:
+		isVirtual, vd, err := tc.virtual.getObjectByName(sc.GetName(), name, objFlags)
+		if isVirtual || vd != nil || err != nil {
+			return haltLookups, vd, err
+		}
+	}
+	return continueLookups, nil, nil
+}
+
+// getNonVirtualDescriptorID looks up a non-virtual descriptor ID by name by
+// going through layers in sequence.
+//
+// All flags except AvoidLeased, RequireMutable and AvoidSynthetic are ignored.
+func (tc *Collection) getNonVirtualDescriptorID(
```

Comment:
This is implemented using the same laddering-style seen in `getDescriptorsByID`.